cmd/crdvalidator/test/e2e/crdvalidator_test.go

```diff
@@ -0,0 +1,150 @@
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/storage/names"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var _ = Describe("crd validation webhook", func() {
+	When("a crd event is emitted", func() {
+		var ctx context.Context
+
+		BeforeEach(func() { ctx = context.Background() })
+		AfterEach(func() { ctx.Done() })
+
+		When("incoming crd event is safe", func() {
+			var crd *apiextensionsv1.CustomResourceDefinition
+
+			BeforeEach(func() {
+				crd = newTestingCRD()
+
+				err := c.Create(ctx, crd)
+				Expect(err).To(BeNil())
+
+				Eventually(func() error {
+					return c.Get(ctx, client.ObjectKeyFromObject(crd), crd)
+				}).Should(Succeed(), "should be able to create a safe crd but was not")
+			})
+
+			AfterEach(func() {
+				By("deleting the testing Bundle resource")
+				Expect(c.Delete(ctx, crd)).To(BeNil())
+			})
+
+			It("should allow the crd update event to occur", func() {
+				Eventually(func() error {
+					if err := c.Get(ctx, client.ObjectKeyFromObject(crd), crd); err != nil {
+						return err
+					}
+
+					crd.Spec.Versions[0].Storage = false
+
+					crd.Spec.Versions = append(crd.Spec.Versions, apiextensionsv1.CustomResourceDefinitionVersion{
+						Name:    "v1alpha2",
+						Served:  true,
+						Storage: true,
+						Schema: &apiextensionsv1.CustomResourceValidation{
+							OpenAPIV3Schema: &apiextensionsv1.JSONSchemaProps{
+								Type:        "object",
+								Description: "my crd schema",
+							},
+						},
+					})
+
+					return c.Update(ctx, crd)
+				}).Should(Succeed())
+			})
+		})
+
+		When("incoming crd event removes a stored version", func() {
+			var crd *apiextensionsv1.CustomResourceDefinition
+
+			BeforeEach(func() {
+				crd = newTestingCRD()
+
+				err := c.Create(ctx, crd)
+				Expect(err).To(BeNil())
+
+				Eventually(func() error {
+					return c.Get(ctx, client.ObjectKeyFromObject(crd), crd)
+				}).Should(Succeed(), "should be able to create a safe crd but was not")
+			})
+
+			AfterEach(func() {
+				By("deleting the testing Bundle resource")
+				Expect(c.Delete(ctx, crd)).To(BeNil())
+			})
+
+			It("should deny admission", func() {
+				Eventually(func() bool {
+					if err := c.Get(ctx, client.ObjectKeyFromObject(crd), crd); err != nil {
+						return false
+					}
+
+					crd.Spec.Versions = []apiextensionsv1.CustomResourceDefinitionVersion{
+						{
+							Name:    "v1alpha2",
+							Served:  true,
+							Storage: true,
+							Schema: &apiextensionsv1.CustomResourceValidation{
+								OpenAPIV3Schema: &apiextensionsv1.JSONSchemaProps{
+									Type:        "object",
+									Description: "my crd schema",
+								},
+							},
+						},
+					}
+
+					err := c.Create(ctx, crd)
+					if err == nil {
+						return false
+					}
+
+					return errorsLooselyEqual(err, errors.New("cannot remove stored versions"))
+				}).Should(BeTrue())
+			})
```

Comment:
A couple of things:

- I'm slightly confused whether we should be firing off an update request vs. a create one
- Can we return an error here and use gomega matchers to enforce the expected error message, e.g. https://github.com/operator-framework/rukpak/blob/bf174efdd76fb100a0d818cbc895f49b7f85b440/test/e2e/plain_provisioner_test.go#L838

Reply:
Yeah so the `Update` here was attempted and continuously failed. If you have the `Create` it succeeds - however, this is not what we should really be expecting and just obfuscated the issue.  Something under-the-hood in Kube is now blocking unsafe storage removals and thus the error that we are expecting is not coming through. This is because the `Update` request does not seem to even hit the webhook and Kube returns its own error. I'm making some changes in an upcoming commit along with a `todo` comment to address this in the future by reevaluating if we need to ensure safe storage removals or if Kube does it by default now.